main/src/ca/uwaterloo/flix/language/phase/Resolver.scala

```diff
@@ -1135,28 +1177,26 @@ object Resolver extends Phase[NamedAst.Root, ResolvedAst.Root] {
   }
 
   /**
-    * Tries to find a sig with the qualified name `qname` in the namespace `ns0`.
+    * Finds the sig with the qualified class `clazz` and ident `sig` in the namespace `ns0`.
     */
-  def tryLookupSig(qname: Name.QName, ns0: Name.NName, root: NamedAst.Root): Option[NamedAst.Sig] = {
-    // Check whether the name is fully-qualified.
-    if (qname.isUnqualified) {
-      // Case 1: Unqualified name. Lookup in the current namespace.
-      val sigOpt = root.sigs.getOrElse(ns0, Map.empty).get(qname.ident.name)
+  def lookupSig(clazz: Name.QName, sig0: Name.Ident, ns0: Name.NName, root: NamedAst.Root): Validation[NamedAst.Sig, ResolutionError] = {
 
-      sigOpt match {
-        case Some(sig) =>
-          // Case 1.2: Found in the current namespace.
-          Some(sig)
-        case None =>
-          // Case 1.1: Try the global namespace.
-          root.sigs.getOrElse(Name.RootNS, Map.empty).get(qname.ident.name)
-      }
-    } else {
-      // Case 2: Qualified. Lookup in the given namespace.
-      root.sigs.getOrElse(qname.namespace, Map.empty).get(qname.ident.name)
+    tryLookupSig(clazz, sig0, ns0, root) match {
+      case Some(sig) => sig.toSuccess
+      case None => ResolutionError.UndefinedSig(clazz, sig0, ns0, sig0.loc).toFailure
     }
   }
 
+  /**
+    * Tries to find a sig with the qualified class `clazz` and ident `sig` in the namespace `ns0`.
+    */
+  def tryLookupSig(clazz: Name.QName, sig: Name.Ident, ns0: Name.NName, root: NamedAst.Root): Option[NamedAst.Sig] = {
+    for {
+      clazz <- tryLookupClass(clazz, ns0, root)
+      result <- clazz.sigs.find(_.sym.name == sig.name)
```

Comment:
I guess this is OK (perf wise) because classes usually contain *very very few* signatures. (Unlike namespaces which can contain hundreds of defs).